Tools v1 0

enricodammers opened the pull request:
Hi All,

I made this pull request so that Kevin can try some things out and work on reformatting the code following his vision.

Test functions are incomplete and i'll add them in a later commit. Probably tomorrow if time works out.

I spend some time implementing Beirle and im quite happy how it works right now.

You should be able to run Beirle for germany without any problem. 

See the email for some examples between one of our inventories and the estimates from Fioletov and Beirle.

Also see the TODO's for things that i'd like to add where possible.

I ran Fioletov at 0.1x0.1 resolution on our supercomputers and the results do not seem to change much compared to the 0.2x0.2 version, only the amout of detail available.

Cheers,

Enrico

kevin-hausmann commented on eocalc/methods/fioletov.py:
```diff
@@ -0,0 +1,448 @@
+        # get nmonths and days
+        dtrange = numpy.array(list(period))
+        ndays = (dtrange.max()-dtrange.min()).days + 1
+        nmonths = len(numpy.unique([f"{day:%Y-%m}" for day in period]))
```
Please remove these lines and use period.days, period.months instead

enricodammers replied:
I propose moving to a datetime or pandas.date_range structure, which is much more intuitive and commonly used.
https://pandas.pydata.org/docs/reference/api/pandas.date_range.html

kevin-hausmann replied:
Yes, I agree in principle, but if you recall, I spent quite a bit of time in the beginning to search for a good solution and we were actually using pandas.date_range back then. There were problem with it, that made me implement DateRange (and I do not take such decisions lightly). Python datetime does not have a period structure. I will keep this in mind, but right now it is only a distraction and would send me on a big detour.

kevin-hausmann commented on eocalc/methods/fioletov.py:
```diff
@@ -0,0 +1,448 @@
+            if split_level <= 25:  # hardcap for now, intended for spatial splitting
+                # splitting by days/months
+                emissions = numpy.zeros((nmonths, len(source_df)), float)
+                bias_grid = numpy.zeros((nmonths, len(lats), len(lons)), float)
```
```suggestion
                bias_grid = numpy.zeros((period.months, len(lats), len(lons)), float)

```

kevin-hausmann commented on eocalc/methods/fioletov.py:
```diff
@@ -0,0 +1,448 @@
+        elif split_operations is True:
+            if split_level <= 25:  # hardcap for now, intended for spatial splitting
+                # splitting by days/months
+                emissions = numpy.zeros((nmonths, len(source_df)), float)
```
```suggestion
                emissions = numpy.zeros((period.months, len(source_df)), float)

```

kevin-hausmann commented on eocalc/methods/fioletov.py:
```diff
@@ -0,0 +1,448 @@
+            obs_to_mem_limit, len(df_obs), len(source_df))
+        # branch
+        # if split_level = 1 / split operations = False, continue like normal
+        if split_operations is False:
```
"No split" (do all on one go) should not be a special case. Please change this code to prepare the splits and then run the same code on each split, even if the number of splits is one.

enricodammers replied:
Ok

enricodammers replied:
i split up the two routes so that a third can be created in the future that allows for splitting by region --> aka splitting a 20x20 region into 4 blocks for example. Here i chose the easier way out for the Germany region, where splitting by month seemed fine. The problem is that for some months there aren't enough observations. What you seem to want is always a split with the value depending on the split_level. Can do that as well, but it will make any correction for diurnal/seasonal cycles and weighting per period a pain.

kevin-hausmann commented on eocalc/methods/fioletov.py:
```diff
@@ -0,0 +1,448 @@
+                bias_grid[:] = numpy.nan
+                # iterate over the months
+                nmonths = numpy.unique([[f"{day:%Y-%m}"] for day in period])
+                months = numpy.unique([[f"{day:%Y-%m}"] for day in period])
```
Duplicate code? Again, please use period.months

kevin-hausmann commented on eocalc/methods/fioletov.py:
```diff
@@ -0,0 +1,448 @@
+                        df_obs_mon, source_df, plumewidth, decay, min_long, min_lat, lats, lons, multiprocessing=multiprocessing)
+                    # test quickplot
+                    # conversion  mol  * m-2 * km2 -> kg/yr =* 1/hr * m2/km2 * kg / mol  * days/year * hours/day =  (kg*m2)/(km2*yr*mol)
+                    source_conversion = (decay * 1e6) * \
```
This is copy&paste from above, please see my comment there

kevin-hausmann commented on eocalc/methods/fioletov.py:
```diff
@@ -0,0 +1,448 @@
+            0.5, grid_lat.ravel()+resolution_lat*0.5
+        grid_polygons = [Polygon(zip([w1, e1, e1, w1, w1], [s1, s1, n1, n1, s1]))
+                         for w1, e1, s1, n1 in zip(wgrid, egrid, sgrid, ngrid)]
+        gp_sources = GeoDataFrame(geometry=grid_polygons)
```
Please check use of the base class _create_grid method

kevin-hausmann commented on eocalc/methods/fioletov.py:
```diff
@@ -0,0 +1,448 @@
+        return df_obs
+
+    @staticmethod
+    def _determine_split_level(obs_to_mem_limit: float, len_obs: int, len_source: int):
```
This whole method is only a one-liner: int(numpy.ceil(len_obs * len_source/obs_to_mem_limit))
If you combine this with my comments on the split operations above, this method is not needed

kevin-hausmann reviewed (changes requested):
I started to work on this code if different ways and will write down notes and suggestions

enricodammers commented on eocalc/methods/beirle.py:
```diff
@@ -0,0 +1,525 @@
+        return pollutant == Pollutant.NO2
+
+    # TODO add plumewidth and decay to parameter pollutant/satellite instrument
+    def run(self, region=MultiPolygon, period=DateRange, pollutant=Pollutant, granularity=TimeAggregation,
```
Fixed.

enricodammers commented:
Trying to push the new Big1 commit, but somehow not sending.

(base) enrico@hyperspec:~/dammerse_tno/space-emissions$ git push -v
Pushing to https://github.com/UBA-DE-Emissionsituation/space-emissions.git
 
Been at this status for a while. I hope it resolves soon, might be the internet here,
MIght be a conflict in this merge or something.


Either way,
I updated all the methods, tests, addid a config file, merged Janot's version into mine, and added the latest improvements.
If anything is missing let me know.
Also comment where you want changes.

I reduced the length of some of the code, but here and there I do not see the point in adding yet another function to reduce the lines by 2 or 3 line.

@jptokaya @kevin-hausmann @Christian-Mielke

kevin-hausmann commented:
Thanks @enricodammers for your work on cleaning all this up!

However, as you can see below, you have a lot of conflicts which need to be resolved before your branch can be merged. Please be careful not to revert changes I made to main over the last few days (see e.g. #45 #44). You will have to rebase your branch on the current main, then resolve your conflicts locally and push the fixed version up here.

Also, I noted that your implementations of the `run()` method do not return the tuple(table, grid) and thus would not work with the webapp.

enricodammers commented:
> h

Morning!, 
I'll rebase the branch and incorporate the latest changes to ensure it works with the webapp.
I knew that I forgot something!
I also split a few functions a bit further in fioletov/beirle

I'll try to resolve it today.

kevin-hausmann commented:
> Morning!, I'll rebase the branch and incorporate the latest changes to ensure it works with the webapp. I knew that I forgot something! I also split a few functions a bit further in fioletov/beirle
> 
> I'll try to resolve it today.

Great! Let me know if you need some help...

enricodammers commented:
@kevin-hausmann @jptokaya @Christian-Mielke 
As far as I can see the rebase went correct. VSC was giving some issues / overlapping changes but those have been resolved.
I also added the latest base.py options, and updated the Calculate* scripts.
Dummy problems have not been updated.

Let me know if any other changes are needed.
